### ISSUE TITLE
Review Performance Issues with Styled Components

### ISSUE THREAD

**Issue by emplums** [primer/components#1039]
In another project at GitHub, @jclem noticed that a `Label` that was rendered from PRC using styled components takes far longer to render than a basic div with inline styles:

> I also think we’re paying a huge cost in styled components. For comparison`, PRC’s <Label /> takes around 1.5ms to render (I’ve seen up to 3ms!). The exact same component styled with plain style={{}} props takes less than 1/10th of a millisecond.

>That’s a small difference until you have 30 rows on screen, each with 3-4 labels.

It looks like this only happens on the first render, and subsequent renders are much faster.

**Comment by jclem** [comment]
**tl;dr** There does appear to be a cost, but I don't think it's appreciable for the majority of use cases, including my current project's at this stage. I would be fine closing this issue @emplums.

---

We're seeing _some_ impact here, but it's very limited because the cost is only seen the first time a given component is rendered (to be clear, that means the first time the component is rendered _anywhere_ in the app). After that, things are as fast as normal.

For example, our first render of [`AvatarStackWrapper`](https://github.com/primer/components/blob/b8b60e72595c36d32a6ecb76f218cac7f21545fa/src/AvatarStack.tsx#L16) always takes longer than expected—around 20ms in development for the first render and 0.1ms subsequently.

![CleanShot 2021-02-10 at 16 55 26](https://user-images.githubusercontent.com/412282/107577849-f435ee00-6bc0-11eb-9acd-d04c55681f05.png)

However, after running a production profile, the impact is even more minimal, at around 3ms for the initial render and 0.1ms for subsequent renders:

![CleanShot 2021-02-10 at 16 57 38](https://user-images.githubusercontent.com/412282/107577986-27787d00-6bc1-11eb-8016-312ed26eef94.png)

I'd have to do a lot more profiling to figure out exactly how much cost there is here, but I don't think it's appreciable for 99% of use cases. It may add up with a page that renders large numbers of _unique_ components, but even then those components would have to have large CSS rule sets (and/or many variations via props passed to the styled component), as it seems this cost grows somewhat with the size of the CSS generated for the styled component. I've only done simple verification of this—removing most of `AvatarStackWrapper`'s CSS brings the development runtime cost of first render down to around 3ms.

To put it in perspective, our first [`Label`](https://github.com/primer/components/blob/b8b60e72595c36d32a6ecb76f218cac7f21545fa/src/Label.tsx#L46), which has a much smaller CSS rule set, is about 2ms.

I'm sure someone with more knowledge of styled components internals would have more insight here, but I'm not currently concerned about the impact.
